メモアプリコマンドライン版のプログラムを作成

sjabcdefin opened the pull request:
クラス(JavaScript)の課題である「メモアプリコマンドライン版」のプログラムを作成いたしました。
ご確認お願いいたします。

保存先は、SQLite のデータベースとしています。
## 事前準備方法
```
# 1. リポジトリをクローンする
git clone -b memo-app <リポジトリのURL>

# 2. クローンしたディレクトリに移動する
cd ./js-practices/04.class

# 3. 必要なパッケージをインストールする
npm install
```

## 使用方法
- メモの追加
1. コマンドラインにて```./main.js```を実行
2. メモの内容を入力
3. テキスト入力後、```Ctrl + D``` で保存

- メモの一覧表示
1. コマンドラインにて```./main.js -l```を実行

- メモの内容表示
1. コマンドラインにて```./main.js -r```を実行
2. 上下矢印キーを使用し、表示したいメモを選択

- メモの削除
1. コマンドラインにて```./main.js -d```を実行
2. 上下矢印キーを使用し、削除したいメモを選択

sjabcdefin committed:
Install sqlite3

memo の保存先をSQLite3のDBとするため、sqlite3をインストール

sjabcdefin committed:
Install enquirer

enquirerをインストール

sjabcdefin committed:
Create a command line app that can list, view, add and delete notes

メモの一覧・参照・追加・編集・削除ができるコマンドラインアプリ(JavaScript)を作成

cafedomancer commented on 04.class/main.js:
Docs の例に合わせて `./memo.js` で動くようにしてください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/7431b8afcedd4657bbc7073a297fa1e157828d71

```./main.js```を```./memo.js```に、```./memo.js```を```memo_model.js```に修正いたしました。
ファイル名変更に伴い、Memo クラスを MemoModel クラスに変更いたしました。

```memo_model.js```について、ファイル名・クラス名を何にするか迷いました。
MVCにおけるModelに該当すると思い、```memo_model.js```, MemoModel クラスに修正しました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,125 @@
+    this.#database = new MemoDatabase("memo.db");
+  }
+
+  async handleOption() {
```
`handleOption` という名前だとオプションの処理だけを行う関数に見えますが、他の実際の処理まで行っているので名前と実態が合ってなさそうです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/535df064bdda89afb7c742f2870ba433f9d44389

```handleOption```を```executeMemoCommand```に修正しました。

このメソッドでは、以下の3つの処理をしており、メモに関するコマンドラインオプションを判定して処理を行っていることから、```executeMemoCommand```に修正いたしました。
1. コマンドラインオプションの判定
2. メモ操作の実行
3. データベース操作（テーブルの作成とデータベースのクローズ）

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,125 @@
+          await this.#addMemo();
+      }
+    } catch (err) {
+      if (err instanceof Error) console.error(`Error: ${err.message}`);
+      else console.error("An unknown error occurred.");
```
if 文はブロックを伴って書いてください。https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Statements/if...else#%E8%A7%A3%E8%AA%AC

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/a1a577c71bd4ebd01aecbf91ba37a98688822fa8

ブロック文を使用するように修正いたしました。

また、以下のように```else```を使用していない箇所があったため、```else```を使用するように修正しました。
```
if (!err) throw new Error("Selecting memo was canceled.");
throw err;
```
↓
```
if (!err) {
   throw new Error("Selecting memo was canceled.");
} else {
   throw err;
}
```

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,125 @@
+    return process.argv.slice(2).length > 1;
+  }
+
+  #isInvalidOption(inputOption) {
```
質問ですが、他の `#hasMultipleOptions` や `#commandLineOption` はメソッドの内部で `process.argv` に直接アクセスしているのに対して、この `#isInvalidOption` はそうなっていないのはなぜでしょうか？ どちらが良い悪いはいったん置いておいて (質問の回答を受けてまたコメントします) 、一貫性がないので質問しています。

sjabcdefin replied:
```#commandLineOption```メソッドと```#isInvalidOption```メソッドでは、コマンドライン引数の1つの要素(```process.argv[2]```)を使用したいため、```process.argv[2]```を2度記載するよりも、変数に一度設定して使いまわした方がよいのではないかと考えました。

```#hasMultipleOptions```メソッド内の```process.argv```については、このメソッド内でしか使用しないため、べた書きしました。

同じ値を使いまわす場合に変数化しており、一貫性については考慮できておりませんでした。🙇‍♀️🙏

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,125 @@
+  }
+
+  async #inputMemo() {
+    const rl = createInterface({
+      input: process.stdin,
+      output: process.stdout,
+    });
```
この設定だとメモをパイプで入力したときに標準出力に出力されます。なぜかを考えた上で、要件にはない挙動なのでそうならないようにしてください。

![CleanShot 2025-03-04 at 07 41 23@2x](https://github.com/user-attachments/assets/ba06bd17-62d8-49a5-8c4c-5c3910abd239)

sjabcdefin replied:
- コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/eeb936fb3b2240b68df207672f1c9db94ec0fade

```output: process.stdin```を設定しているため、入力した内容がそのまま標準出力されてしまっていました。
```output: process.stdin```を削除しました。

- コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/c83b6a6086b1aaa7278f1079f6fcc733253f0041

```output:```の削除により、メモ入力中に、```Ctrl+C```が押下された場合、```SIGINT```シグナルが Node.js に送られ、メモが保存されずに、プロセスが即時終了するようになった認識です。
また、メモ選択中に、```Ctrl+C```が押下された場合、例外をスローしていましたが、```Ctrl+C```押下は、ユーザーの意図的な操作であるため、エラーとして扱うのではなく、適切にクリーンアップしてプロセスを終了したほうがよいと考えました。
そのため、メモ入力中及び選択中に、```Ctrl+C```が押下された場合、ログの出力 → データベースのクローズ → プロセスの終了 という同じ処理が実行されるよう修正いたしました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,125 @@
+    });
+    let memoContent = "";
+    try {
+      for await (const line of rl) {
+        memoContent += line + "\n";
+      }
+      return memoContent.trimEnd();
```
そもそも `trimEnd` しなくていいような実装を考えてください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/f839cb79d8d801bd54b804c7f3cb60946b4fb107

```memoContents```配列に、各行を格納し、```join("\n")```で改行を追加するように修正いたしました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,125 @@
+    }
+  }
+
+  async #selectMemo(action = "see") {
```
`"see"` も `"delete"` も対等な処理なので `"see"` をデフォルトにしている理由がよく分からないです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/24c36458d6236c835cbe8a0da2769cf7db27bdb7

すみません。
とりあえず、```"see"```をデフォルトにしておくかという安易な考えでした。
デフォルト引数を指定せず、```"see"```、```"delete"```それぞれを引数として渡すように修正しました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,125 @@
+      ];
+      return await prompt(question);
+    } catch (err) {
+      if (!err) throw new Error("Selecting memo was canceled.");
```
`!err` だと例外として何を期待しているのかが分からないので、値を指定して条件を明示的にしてください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/6e83ad632d9a6457aa70a4aba97bcf43145c0f3f

```if (err === "")```に修正いたしました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,125 @@
+
+  async #fetchMemos() {
+    const rows = await new Memo(this.#database).fetchAll();
+    return rows.map((row) => ({
+      name: row.content.split("\n")[0],
+      value: row.id,
+    }));
```
`{ name: ..., value: ... }` という形のオブジェクトは enquirer に特化しているものです。それを得るメソッドに汎用的な名前である `fetchMemos` というのを付けるのはちょっと無理がある気がします。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/7fede584e0d82809013d68caecfa518773bddd18

enquirer を使用し、プロンプト表示用のメモデータを取得するメソッドであるため、```fetchMemos```から```fetchMemosForEnquirerPrompt```に変更いたしました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,125 @@
+    return option;
+  }
+
+  async #inputMemo() {
```
`#inputMemo` という名前なのに返しているのはメモそのものではなくメモの中身になっています。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/3c89b87e834937398a51d3ca1325782e52a23973

```inputMemo```から```readMemoContentFromInput```に変更いたしました。
メモの中身の入力だけではなく、メモの中身を返しているため、```inputMemoContent```より```readMemoContentFromInput```の方が、メソッド名として適していると考えました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,125 @@
+
+  async #displayMemos() {
+    const memos = await this.#fetchMemos();
+    memos.forEach((memo) => console.log(memo.name));
```
戻り値のないアロー関数にはブロック文体を使ってください。戻り値がないことを明示するためです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/ca4f886ae6ec52cf548e877189492f8a2fafda93

ブロック文体を使用するよう修正いたしました。
```
memos.forEach((memo) => {
  console.log(memo.name);
});
```

cafedomancer commented on 04.class/memo_database.js:
```diff
@@ -0,0 +1,48 @@
+import sqlite3 from "sqlite3";
+import { runQuery, allQuery, getQuery, closeDatabase } from "./sqlite_utils.js";
+
+const createTableQuery = `
+CREATE TABLE IF NOT EXISTS memos (
+  id INTEGER PRIMARY KEY AUTOINCREMENT,
+  content TEXT NOT NULL
+)
+`;
```
この書き方だと先頭と末尾に余分な空行が含まれます。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/9de5f6497c072b9500b942723142075db1ab4c0d

不要な改行とインデントを削除し、一行に統一いたしました。

cafedomancer commented on 04.class/memo_database.js:
```diff
@@ -0,0 +1,48 @@
+  content TEXT NOT NULL
+)
+`;
+const insertMemoQuery = `INSERT INTO memos (content) VALUES (?)`;
```
文字列中に変数を埋め込んでいないのにテンプレート文字列が使われています。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/e1c1d2f982330b0eb0531f6c3699da63a6ac09d9

テンプレート文字列の代わりに通常の文字列を使用するように修正いたしました。
```insertMemoQuery```だけでなく、```createTableQuery```, ```selectAllMemoQuery```, ```selectMemoQuery```, ```deleteMemoQuery```も文字列中に変数を埋め込んでいないため、通常の文字列に修正いたしました。

cafedomancer commented on 04.class/memo_database.js:
```diff
@@ -0,0 +1,48 @@
+)
+`;
+const insertMemoQuery = `INSERT INTO memos (content) VALUES (?)`;
+const selectAllMemoQuery = `SELECT * FROM memos`;
```
このクエリだとメモの順序が不定になります。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/032ae735b23f4d5951a1612a6b3bcffa03367815

```ORDER BY```句を使用し、表示順が一定（メモIDが昇順）になるように修正いたしました。

cafedomancer commented on 04.class/memo_database.js:
```diff
@@ -0,0 +1,48 @@
+`;
+const insertMemoQuery = `INSERT INTO memos (content) VALUES (?)`;
+const selectAllMemoQuery = `SELECT * FROM memos`;
+const selectMemoQuery = `SELECT * FROM memos WHERE id = (?)`;
```
`id = ...` の部分に括弧はいらないような。INSERT 文では必要ですが。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/15f8f603b777749fd885c2292b997070374cdef6

```selectMemoQuery```と```deleteMemoQuery```のプレースホルダの不要な括弧を削除いたしました。
（```id = ...```は、単一の値が設定されるため、括弧は不要）

cafedomancer commented on 04.class/memo_database.js:
```diff
@@ -0,0 +1,48 @@
+    return await allQuery(this.#db, selectAllMemoQuery);
+  }
+
+  async getById(memoID) {
```
質問ですが、`add` メソッドの `content` には memo がついていないのに、この `getById` メソッドの `memoID` にはなぜ memo がついているんでしょうか？

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/9f61c84baca6dbe2831071b7daf6696527448e20

失礼いたしました。一貫性について意識が足りませんでした。
```memoID```を```id```に修正いたしました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,53 @@
+class Memo {
+  #database;
```
このクラスが何を表しているかによりますが、それぞれの単体のメモを表しているのなら、それがデータベースを必要とするのはおかしいです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/d003192de0cc4bba098418c5e6aefc9605aff3ad

ご認識の通り、Memo クラス(修正後は MemoModel クラス)は、単体のメモを表しています。
MemoModel クラスのフィールドから```#database```を削除しました。
MemoModel クラスの```save```, ```fetchAll```, ```fetchById```, ```delete``` メソッドで、MemoDatabaseインスタンスをメソッド引数として受け取るように修正しました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,125 @@
+  }
+
+  async #fetchMemos() {
+    const rows = await new Memo(this.#database).fetchAll();
+    return rows.map((row) => ({
+      name: row.content.split("\n")[0],
+      value: row.id,
+    }));
```
`fetchAll()` で返ってくるのは row ではなく memo だと思いますが。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/f6b7300a7040d69370e4f96bda5bc49cc39d2b84

以下のように修正いたしました。
```
const memos = await new MemoModel().fetchAll(this.#database);
return memos.map((memo) => ({
  name: memo.content.split("\n")[0],
  value: memo.id,
}));
```

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,53 @@
+  #id;
+  #content;
+
+  constructor(memoDB, id = null, content = "") {
```
渡さなくていい引数があるのなら、そもそも渡さなくてもいいような実装にしてください。普通の引数が使われていて、かつ渡す必要がないのに引数の仕組み上渡さなければいけないから `undefined` などが渡されていると、呼び出し側だけ見ても何のために渡しているのか、それが何の値なのかが分からないので読みづらいです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/f5a73052602867dd034ad926395b47c62f082013

分割代入構文を使用し、渡す必要がない引数については、渡さないように修正いたしました。

cafedomancer commented on 04.class/memo.js:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  #isContentEmpty() {
+    return !this.#content;
```
何をチェックしているのか分からないので条件を明示してください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/20c281363aab5d7e03923b2d89ba4552653757d4

条件を明示いたしました。
```
#isContentEmpty() {
  return this.#content === "";
}
```

sjabcdefin committed:
Rename file and class

- Docs の例に合わせてファイル名を変更(main.js → memo.js)
- 上記修正に伴い、memo.js → memo_model.js に修正
- ファイル名の修正に伴い、ファイル名に合わせる形でクラス名を修正(Memo → MemoModel)

sjabcdefin committed:
Refactor Memo class to remove direct dependency on database

Memo クラスが個々のメモを表しているのに、database をプロパティとして持つのは不自然であるため

sjabcdefin committed:
Fix the Memo class constructor argument to destructuring assignment notation

可読性向上のため、Memo クラスのコンストラクタの引数を分割代入表記にし、必要な引数のみを引き渡せるように修正

sjabcdefin committed:
Rename handleOption to executeMemoCommand

実際の処理と関数名の実態を合わせるため、メソッド名を変更

sjabcdefin committed:
Fix to use block style when using if statement

if文使用時は、常にブロック文({ ... })を使用することが推奨されているため

sjabcdefin committed:
Remove output option from readline interface

入力内容が自動的に標準出力に表示されないようにするため

sjabcdefin committed:
Add memo.db to .gitignore

memo.db は git 管理する必要がないため

sjabcdefin committed:
Refactor #inputMemo method to eliminate the trimEnd()

trimEnd() を使用せず、末尾の不要な改行を避ける実装に変更するため

sjabcdefin committed:
Refactor #selectMemo method to not use default argument

"see" も "delete" も対等な処理であり、どちらかをデフォルト引数として指定しているのは不自然なため

sjabcdefin committed:
Rename #fetchMemos to #fetchMemosForEnquirerPrompt

{ name: ..., value: ... }という形のオブジェクトは enquirer に特化しており、それを得るメソッドに汎用的な名前である fetchMemos を付けるのは違和感があるため

sjabcdefin committed:
Rename #inputMemo from #readMemoContentFromInput

メモの中身の入力だけではなく、メモの中身を返しているため、inputMemoContent より readMemoContentFromInput の方がメソッド名として適しているため

sjabcdefin committed:
Fix to use block style for Arrow function with no return value

戻り値がないことを明示するため

sjabcdefin committed:
Remove unnecessary line breaks and indentation from createTableQuery

createTableQueryから不要な改行とインデントを削除し、一行に統一

sjabcdefin committed:
Fix the SQL query to use a regular string instead of a template string

文字列中に変数を埋め込んでいないため

sjabcdefin committed:
Fix selectAllMemoQuery to use ORDER BY clause

メモ一覧を取得する際に、表示順が一定になるようにするため

sjabcdefin committed:
Remove unnecessary parentheses from selectMemoQuery and deleteMemoQuery

selectMemoQuery と deleteMemoQueryのプレースホルダ (?) に不要な括弧が含まれていたため

sjabcdefin committed:
Rename "memoID" to "id" in getById and delete method

変数の命名に一貫性がなかったため

sjabcdefin committed:
Rename "rows" to "memos" in #fetchMemosForEnquirerPrompt method

fetchAll()で返ってくるのは row ではなく、memo であるため

sjabcdefin committed:
Clarify error handling when selecting a memo is canceled

!err だと例外として何を期待しているか不明なため

sjabcdefin committed:
Clarify condition in #isContentEmpty method

メモ内容が空であることを明示的に確認するため

sjabcdefin committed:
Fix behavior when processing Ctrl+C

メモ入力時および選択時のCtrl+C 押下時の挙動を合わせるため

sjabcdefin commented:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/56dcfca3a49288e07a0584e885bdaeea4f51c219

memo.db は git 管理不要なため、.gitignore に追加いたしました。

cafedomancer commented on 04.class/sqlite_utils.js:
直接実行されないファイルなのに実行権限がついていそうです。

sjabcdefin replied:
sqlite_utils.js について、実行権限を削除いたしました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,143 @@
+    });
+  }
+
+  async executeMemoCommand() {
```
`MempApp().executeMemoCommand()` の部分で memo が重複しているように思います。あと、コマンドを実行しているのはシェルの `$ ./memo.js` という操作であって、このメソッドではないような気がします。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/a7ebdb4858ba68f1bf566d407bab46f585ac5bf1

コマンドライン入力を処理するという表現となるよう、```executeMemoCommand()```を```processCommandLine()```に変更いたしました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,143 @@
+    return !options.includes(inputOption);
+  }
+
+  #commandLineOption() {
```
メソッド名は動詞始まりにしてください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/09f7b9069ffdb05f47b44e50280b80a665416c5d

```#commandLineOption()```を```#parseCommandLineOption()```に変更いたしました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,143 @@
+    const rl = createInterface({
+      input: process.stdin,
+    });
+    let memoContents = [];
```
再代入されない変数は `const` を使って宣言してください。あと、プログラム中において、メモの contents と content の区別がよく分からないです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/6fd6e49be5f9584cae0c2473b2342d55049ffe7b

再代入されない変数について、```let```ではなく、```const```で宣言するように修正いたしました。

コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/f1d51fdae1501462a00bdf1a26239b98c08d4dc6

変数名```memoContents```は、メモの内容全体(1行単位)を表すために使用していましたが、
他の変数```content```は、メモの内容全体を表しており(```memoContents```配列を文字列として結合した値)、変数名に一貫性がありませんでした。

```memoContents```は、ユーザが入力した1行ごとの内容（行単位）の集合であるため、```inputLines```に修正いたしました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,143 @@
+  async #selectMemo(action) {
+    const memos = await this.#fetchMemosForEnquirerPrompt();
+    try {
+      const { prompt } = enquirer;
```
`prompt` を分割代入して使う必要は別にないと思います。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/128acfbc59a44a1d4f903679e8b118a92e339797

```enquirer.prompt```に修正いたしました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,143 @@
+    }
+  }
+
+  async #fetchMemosForEnquirerPrompt() {
+    const memos = await new MemoModel().fetchAll(this.#database);
+    return memos.map((memo) => ({
+      name: memo.content.split("\n")[0],
+      value: memo.id,
+    }));
+  }
```
enquirer に関する処理は `#selectMemo` で行っているのだから、この処理もそこに閉じているべきだと思います。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/3d3ced4f36d069a48cf4fd9d1bd0e5bd11b3af83

https://github.com/sjabcdefin/js-practices/pull/4#discussion_r2005549214 の修正とあわせて修正いたしました。

メモ全体を取得する処理は、```#fetchMemos()```で実施し、enquirer で使用する選択肢用配列は、```#selectMemo()```内で生成するように修正いたしました。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,143 @@
+    }
+  }
+
+  async #selectMemo(action) {
```
`#selectMemo` という名前なのに返しているのはメモの ID です。メモのオブジェクトを返すようにしたほうが取り回しもよくなるでしょう。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/a0e7d29cafe13c6e656cd4a00b6e469465d2370b

https://github.com/sjabcdefin/js-practices/pull/4#discussion_r2005539069 の修正とあわせて修正しております。

メモのオブジェクトを返すように修正いたしました。
この修正に伴い、メモIDからメモの内容を取得するデータベース操作が不要となったため、memo_database.js の ```getById()```、sqlite_utils.js の ```getQuery()``` を削除いたしました。

cafedomancer commented on 04.class/memo_model.js:
```diff
@@ -0,0 +1,58 @@
+    return this.#content === "";
+  }
+
+  async save(database) {
```
https://github.com/sjabcdefin/js-practices/pull/4#discussion_r1978355405 にコメントしている問題が解決していないです。`MemoModel` がメモを表すクラスなのであれば、それがデータベースに関する処理を行っているのはおかしいし、データベースに依存しているのもおかしいと言っています。Rails の Active Record の実装に引っ張られていませんか？

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/a0e7d29cafe13c6e656cd4a00b6e469465d2370b

https://github.com/sjabcdefin/js-practices/pull/4#discussion_r2005535383 の修正とあわせて修正しております。

MemoModelクラス内のデータベース関連の処理を削除し、データの保持のみをするように修正いたしました。
データベース関連の処理は、MemoAppクラスにて実行するようにしております。

cafedomancer commented on 04.class/memo_model.js:
```diff
@@ -0,0 +1,58 @@
+    this.#content = memoContent;
+  }
+
+  #isContentEmpty() {
+    return this.#content === "";
+  }
+
+  async save(database) {
+    if (this.#isContentEmpty()) {
+      throw new Error("Memo content cannot be empty. Please enter some text.");
+    }
+    await database.add(this.content);
+  }
+
+  async fetchAll(database) {
+    const rows = await database.getAll();
+    if (!rows.length) {
+      throw new Error(
+        "No memos available. Use the app without options to add a new memo.",
+      );
+    }
+    return rows.map(
+      (row) => new MemoModel({ id: row.id, content: row.content }),
+    );
+  }
+
+  async fetchById(database) {
+    const row = await database.getById(this.id);
+    this.content = row.content;
+  }
+
+  async delete(database) {
+    await database.delete(this.id);
+  }
```
public メソッド、private メソッドの順に定義してください。

sjabcdefin replied:
https://github.com/sjabcdefin/js-practices/pull/4#discussion_r2005539069 の修正に伴い、MemoModelクラスは、publicメソッドのみとなったため、修正しておりません。
今後、public メソッド、private メソッドの順に定義するようにいたします。

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,143 @@
+  }
+
+  async #displayMemos() {
+    const memos = await this.#fetchMemosForEnquirerPrompt();
```
enquirer を使っていないのになんで `#fetchMemosForEnquirerPrompt()` が呼び出されているんでしょうか...。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/3d3ced4f36d069a48cf4fd9d1bd0e5bd11b3af83

https://github.com/sjabcdefin/js-practices/pull/4#discussion_r2005546096 の修正とあわせて修正しております。

```#fetchMemosForEnquirerPrompt() ```を```#fetchMemos()```に変更し、MemoModelオブジェクトの集合を返すように修正いたしました。
また、可読性向上のため、MemoModelオブジェクトに、メモの内容の1行目を表す```title```を追加しました。

cafedomancer commented on 04.class/memo_database.js:
```diff
@@ -0,0 +1,44 @@
+    await runQuery(this.#db, createTableQuery);
+  }
+
+  async add(content) {
+    return await runQuery(this.#db, insertMemoQuery, [content]);
+  }
+
+  async getAll() {
+    return await allQuery(this.#db, selectAllMemoQuery);
+  }
+
+  async getById(id) {
+    return await getQuery(this.#db, selectMemoQuery, [id]);
+  }
```
データベースに関する処理を扱うのなら、データベースの操作に直接的に対応する名前 (INSERT など) にしたほうがよいと思います。間接的に対応する名前になっていると、他のことをするのかどうかの区別がいまいち付かないからです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/cd7dc1925ed8e8f73dbfc0baa9d56f2f37d6b0ef

```add()```を```insert()```に、```getAll()```を```selectAll()```に変更いたしました。
なお、```getById```メソッドは、https://github.com/sjabcdefin/js-practices/pull/4#discussion_r2005535383 の修正に伴い、削除いたしました。

sjabcdefin committed:
Remove execute permission for sqlite_utils.js

直接実行されないファイルであり、実行権限が不要なため

sjabcdefin committed:
Refactor entire memo app to separete database logic from MemoModel and return memo objects after selecting memo

MemoModelをデータの表現に専念させるため、MemoModelクラス内のデータベース関連の処理を削除（MemoAppクラスにて、データベース関連の処理を実行するように修正）
取り回しを容易にするため、selectMemoメソッドの戻り値として、memoIdではなく、メモオブジェクトを返すように修正
上記の修正に伴い、不要となった、メソッドをmemo_database.jsやsqlite_utils.jsから削除

sjabcdefin committed:
Rename fetchMemosForEnquirerPrompt to fetchMemos and Refactor fetchMemos and selectMemo method

- 汎用性向上のため、fetchMemosメソッドは、メモオブジェクトを返す処理に修正
- enquirerの処理をまとめるため、プロンプト選択肢用のメモオブジェクトは、selectMemoメソッド内で生成するように修正
- 可読性向上のため、MemoModelクラスに、メモの内容の1行目を保持するtitle変数を設定

sjabcdefin committed:
Rename executeMemoCommand to processCommandLine

メソッド内の処理をより明確に表すため

sjabcdefin committed:
Rename commandLineOption to parseCommandLineOption

メソッド名を動詞始まりにするため

sjabcdefin committed:
Rename memoContents to inputLines in readMemoContentFromInput method

より適切な変数名にするため

sjabcdefin committed:
Change the inputLines declarator from let to const

inputLinesは、再代入がない変数なため

sjabcdefin committed:
Change from destructing assignment to direct assignment

分割代入する意味がないため

sjabcdefin committed:
Rename add to insert and getAll to selectAll

データベースに関する処理を行うため、データベース操作に直接的に対応する名前に変更

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,150 @@
+    });
+  }
+
+  async processCommandLine() {
```
コマンドラインを処理するというのはちょっと大きく出過ぎた名前な気がしますね...。逆に言うと何をするメソッドなのかが分からないです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/2c9b7ba62836b4721545bdb884ee4e958b7e6b5a

```runOperationBasedOnOption```に修正いたしました。
```./memo.js```では、```new MemoApp.runOperationBasedOnOption();```として、オプションに応じたメモアプリの処理を実行することがメソッド名から明確に読み取れるようにしました。

cafedomancer commented on 04.class/memo_database.js:
```diff
@@ -0,0 +1,39 @@
+import sqlite3 from "sqlite3";
+import { runQuery, allQuery, closeDatabase } from "./sqlite_utils.js";
+
+const createTableQuery =
+  "CREATE TABLE IF NOT EXISTS memos (id INTEGER PRIMARY KEY AUTOINCREMENT, content TEXT NOT NULL)";
+const insertMemoQuery = "INSERT INTO memos (content) VALUES (?)";
+const selectAllMemoQuery = "SELECT * FROM memos ORDER BY id";
+const deleteMemoQuery = "DELETE FROM memos WHERE id = ?";
```
コメントし忘れていましたが、これらはトップレベルではなく `MemoDatabase` クラスの中で定義されているべき内容だと思います。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/1e6afdf973c84ca0bf014117ddd9617de69d1ffc

```MemoDatabase```クラスの中で定義するように修正いたしました。

cafedomancer commented on 04.class/memo_model.js:
```diff
@@ -0,0 +1,36 @@
+
+  constructor(id, content) {
+    this.id = id;
+    this.content = content;
+    this.title = content;
```
`this.content` と `this.title` のどちらにも `content` が入っていて区別がつかないです。また、仕様としては 、メモの参照時に各メモの 1 行目を扱うというだけで、1 行目がメモのタイトルであるかどうかは分からない気がします。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/579996a0528491cb77f3bcd791f7bde60b20e721

メモの参照時に各メモの1行目を扱うための変数であるため、```title```を```firstLineOfContent```に修正いたしました。
また、```this.title = content;```を削除し、```content```のセッターの中で、```firstLineOfContent```にも設定するように修正いたしました。 
```
  set content(memoContent) {
    this.#content = memoContent;
    this.#firstLineOfContent = memoContent.split("\n")[0];
  }
```

cafedomancer commented on 04.class/memo_model.js:
```diff
@@ -0,0 +1,36 @@
+    return this.#title;
+  }
+
+  set id(memoId) {
```
引数名 `id` でいいと思います。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/631c18827cfc738993c1105529a94f5961cfabde

```set id(memoId)``` の引数名を```id```に修正いたしました。

cafedomancer commented on 04.class/memo_model.js:
```diff
@@ -0,0 +1,36 @@
+    return this.#title;
+  }
+
+  set id(memoId) {
+    this.#id = memoId;
+  }
+  set content(memoContent) {
+    this.#content = memoContent;
+  }
```
メソッドの間に空行がないです。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/a8d54fca5d3755e437832e590d3a4f10d0c967c7

メソッド間に空行を追加いたしました。

sjabcdefin committed:
Rename processCommandLine to runOperationBasedOnOption

より適切なメソッド名称にするため（processCommandLineだと何をするメソッドか分からないため)

sjabcdefin committed:
Fix to define Query variables  in MemoDatabase class

クエリは、一般的にクラス内に定義されるべきであるため

sjabcdefin committed:
Refactor MemoModel to use firstLineOfContent instead of title for the first line of the content

仕様にあった実装にするため

sjabcdefin committed:
Rename the argument name of the id setter from memoId to id

一貫性を保つため

sjabcdefin committed:
Add a line break between methods

可読性向上のため

cafedomancer commented on 04.class/memo_app.js:
```diff
@@ -0,0 +1,150 @@
+    });
+  }
+
+  async runOperationBasedOnOption() {
```
単に `run` でいいんじゃないでしょうか。メモアプリの単なるエントリーポイントなので、凝った名前をつける必要はないと思います。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/3e84542e4094137d689fd56944f794735e0ced64

複雑に考えすぎていました。😵
```run```に修正いたしました。

cafedomancer commented on 04.class/memo_database.js:
```diff
@@ -0,0 +1,39 @@
+import { runQuery, allQuery, closeDatabase } from "./sqlite_utils.js";
+
+class MemoDatabase {
+  static createTableQuery =
+    "CREATE TABLE IF NOT EXISTS memos (id INTEGER PRIMARY KEY AUTOINCREMENT, content TEXT NOT NULL)";
+  static insertMemoQuery = "INSERT INTO memos (content) VALUES (?)";
+  static selectAllMemoQuery = "SELECT * FROM memos ORDER BY id";
+  static deleteMemoQuery = "DELETE FROM memos WHERE id = ?";
```
クラスの外から参照されないプロパティはプライベートにしてください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/a19282778152bce3c7bbe558ee2ce2b1acd9d5f9

クラス外から参照されないQuery変数について、プライベートに修正いたしました。

cafedomancer commented on 04.class/memo_model.js:
```diff
@@ -0,0 +1,33 @@
+    this.#id = id;
+  }
+
+  set content(memoContent) {
```
`memoContent` も直してください。https://github.com/sjabcdefin/js-practices/pull/4#discussion_r2007394833

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/5545321750ff50234fc4a469eb44e7d21338bf1a

すみません。前回指摘された際に修正できた内容でした。
```content```に修正いたしました。

cafedomancer commented on 04.class/memo_model.js:
```diff
@@ -0,0 +1,33 @@
+
+  set content(memoContent) {
+    this.#content = memoContent;
+    this.#firstLineOfContent = memoContent.split("\n")[0];
```
別のプロパティから導出できるものを新たなプロパティとして持つ必要はないです。`firstLineOfContent` の getter だけ定義して、外部からはそのプロパティがあるように見せ、内部的には `content` から導出するようにしてください。

sjabcdefin replied:
コミット：https://github.com/sjabcdefin/js-practices/pull/4/commits/8b0964fbf4bd43b396d9aced47f2ce91c3e0e848

`firstLineOfContent`の getter のみを定義するように修正いたしました。
```
get firstLineOfContent() {
  return this.#content.split("\n")[0];
}
```

sjabcdefin committed:
Rename runOperationBasedOnOption to run

メモアプリのエントリーポイントであるため、単純なメソッド名とする

sjabcdefin committed:
Fix the query variables to private variables

クラス外からの参照がないため

sjabcdefin committed:
Rename the argument name of the content setter from memoContent to content

一貫性を保つため

sjabcdefin committed:
Fix to define only the getter of firstLineOfContent

別のプロパティから導出できるものを新たなプロパティとして持つ必要はないため
